### ISSUE TITLE
Capturing method call arguments with separate API method

### ISSUE THREAD

**Issue by NagRock** [NagRock/ts-mockito#19]
in https://github.com/NagRock/ts-mockito/issues/16 were proposed fluent API for capturing arguments and setting return value. But this API is in conflict with https://github.com/NagRock/ts-mockito/issues/12
So after conversation with @michalstocki we thought about such solutions:

1. `when(someObject.doSomething(anything())).thenReturn(3).thenCapture(captor);` - but after implementing https://github.com/NagRock/ts-mockito/issues/12 this API would be illogical because we will be able to define more returning values with fluent API e.g. `when(someObject.doSomething(anything())).thenReturn(3).thenReturn(4).thenReturn(5).thenCapture(captor);` - thenReturn call order will have matter. So thenCapture does not fit here.
2. `verify(someObject.doSomething(captor.capture()));` - similar to JavaMockito version, but we decided that using `verify` API function is not best solution, maybe this will be added as option to be more compatible with Java version
3. `when(someObject.doSomething(anything())).thenReturnAndCapture(3, captor)` - this not looks nice, and arguments type checking will be impossible.
4. `when(someObject.doSomething(anything(firstCaptor))).thenReturn(3);` - on first sight looks nice, but capturing arguments should not have any conditions while here we can define e.g. `between(3, 4, capture)` - and this looks bad
5. `capture(someObject.doSomething(captor));` - this looks best at all, we do not use conditions, we strictly says that we want to capture arguments

**Comment by michalstocki** [comment]
@NagRock thanks for great summary.

I have yet another idea that might be considered while introducing the `capture()` method. Definitely the most common usage of the capturing mechanism is to get the value of the most recent call. It states as follows assuming that we're going to implement the 5th solution:
```typescript
const firstArgCaptor = new Captor();
const secondArgCaptor = new Captor();
capture(someObject.doSomething(firstArgCaptor, secondArgCaptor));

// ... some logic invoking someObject.doSomething(...)

const firstArgValue = firstArgCaptor.getLastCallValue();
const secondArgValue =  secondArgCaptor.getLastCallValue();
expect(firstArgValue).to.equal(expectedFirstArg);
// do something with the captured value:
secondArgValue();
```
As we can see, the syntax is quite verbose. I wish, I could do this within a several lines, for example:
```typescript
const captorApplier = capture(someObject.doSomething(new Captor(), new Captor()));

// ... some logic invoking someObject.doSomething(...)

const [firstArgValue, secondArgValue] = captorApplier.getAll().map(c => c.getLastCallValue());
expect(firstArgValue).to.equal(expectedFirstArg);
secondArgValue();
```
It could be even shorter if the `capture()` could be called after invocation of the method (as `verify()` in Java Mockito), and assuming that the `getLastCallValues()` will map captors into it's values:
```typescript
// ... some logic invoking someObject.doSomething(...)

const [firstArgValue, secondArgValue] = capture(someObject.doSomething(new Captor(), new Captor())).getLastCallValues();
expect(firstArgValue).to.equal(expectedFirstArg);
secondArgValue();
```

**Comment by NagRock** [comment]
@michalstocki great idea! I've already implemented it on branch https://github.com/NagRock/ts-mockito/tree/issue/12

Syntax will be like you proposed:
`[firstArg, secondArg] = capture(someObject.doSomething).last();`

And then you will get two typed values `firstArg` and `secondArg`
And you will be able to get: `first()`, `second()`, `third()`, `beforeLast()`, `last()`, `byCallIndex(5)`

**Comment by michalstocki** [comment]
Wonderful! I like these short methods `first()`, `last()`. 
It's especially handy, that TypeScript is able to determine the argument types:
```
export class Bar {
    public doSomething(a:string, b:Bar, c:number):string {
        return 'done';
    }
}

const b:Bar = new Bar();
const arr = capture(b.doSomething).last();
arr[0] // TypeScript: string!
arr[0].charAt(0); // ok
arr[1] // TypeScript: Bar
```

**Comment by NagRock** [comment]
Should be fixed in [2.0 version](https://github.com/NagRock/ts-mockito/releases/tag/v2.0.0). Please reopen if something is still wrong.
